### PR TITLE
Rule S4057: Locales should be set for data types

### LINKED ISSUE #511: Rule S4057: Locales should be set for data types
[RSPEC-4057](https://jira.sonarsource.com/browse/RSPEC-4057)

When you create a `DataTable` or `DataSet`, you should set the locale explicitly. By default, the locale for these types is the current culture. For data that is stored in a database or file and is shared globally, the locale should ordinarily be set to the invariant culture (`CultureInfo.InvariantCulture`).

Contributes to [MMF-963](https://jira.sonarsource.com/browse/MMF-963).

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #511

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> invalidTypes = new HashSet<KnownType>
+        {
+            KnownType.System_Data_DataTable,
+            KnownType.System_Data_DataSet
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 27 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![CRITICAL][CRITICAL] 1 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 27 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![CRITICAL][CRITICAL] 1 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 27 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![CRITICAL][CRITICAL] 1 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Vsix/source.extension.vsixmanifest [hunk 2]
```diff
@@ -25,5 +25,5 @@
     <Asset Type="Microsoft.VisualStudio.Analyzer" d:Source="Project" d:ProjectName="SonarAnalyzer.VisualBasic" Path="|SonarAnalyzer.VisualBasic|" />
     <Asset Type="Microsoft.VisualStudio.MefComponent" d:Source="Project" d:ProjectName="SonarAnalyzer.VisualBasic" Path="|SonarAnalyzer.VisualBasic|" />
     <Asset Type="Microsoft.VisualStudio.Analyzer" d:Source="File" Path="Google.Protobuf.dll" />
-  </Assets>
+  </Assets>
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Vsix/source.extension.vsixmanifest]: Revert the change to this file.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+                if (memberAccess.Expression == null)
+                {
+                    return null;
+                }
+
+                return model.GetSymbolInfo(memberAccess.Expression).Symbol;
+            }
+
+            var conditionalAccess = variable as ConditionalAccessExpressionSyntax;
+            if (conditionalAccess != null)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: That would be simpler:

```
return conditionalAccess?.Expression != null
    ? model.GetSymbolInfo(conditionalAccess.Expression).Symbol
    : null;
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4057";
+        private const string MessageFormat = "Set the locale for this '{0}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> invalidTypes = new HashSet<KnownType>
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: These types are not invalid. Rename to something like "dataTypes" or "checkedTypes"?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
+using System;
+using System.Data;
+using System.Globalization;
+
+namespace Tests.Diagnostics
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.cs]: Add more tests:
1. Create DataTable and sets the Locale property on some different object.
2. Create two DataTable/Sets, but set the Locale property only on one of them.
3. DataTable dt; dt = new DataTable(); dt.Locale=...

Should we to check if the culture is set to invariant (I presume not)?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                            }
+                        }, SyntaxKind.ObjectCreationExpression);
+
+                    compilationStartContext.RegisterSyntaxNodeActionInNonGenerated(
+                        c =>
+                        {
+                            var assignmentExpression = (AssignmentExpressionSyntax)c.Node;
+                            var propertySymbol = GetPropertySymbol(assignmentExpression, c.SemanticModel);
+
+                            if (propertySymbol == null ||
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: [Optional] perhaps reverse condition and put the code below inside if?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+                    .FirstAncestorOrSelf<ObjectCreationExpressionSyntax>()
+                    ?.FirstAncestorOrSelf<AssignmentExpressionSyntax>()
+                    ?.Left;
+                if (leftSideOfParentAssignment != null)
+                {
+                    return model.GetSymbolInfo(leftSideOfParentAssignment).Symbol;
+                }
+
+                var lastVariable = identifier.FirstAncestorOrSelf<VariableDeclarationSyntax>()?.Variables.LastOrDefault();
+                if (lastVariable == null)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: simpler:

`return lastVariable != null ? model.GetDeclaredSymbol(lastVariable) : null;`

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+                                }
+
+                                c.ReportDiagnosticIfNonGenerated(Diagnostic.Create(rule,
+                                    invalidCreation.Syntax.GetLocation(), typeName), c.Compilation);
+                            }
+                        });
+                });
+        }
+
+        private static SyntaxNode GetAssignedToSyntax(ObjectCreationExpressionSyntax objectCreation)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: Perhaps better as GetAssignmentTargetVariable?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+                    compilationStartContext.RegisterCompilationEndAction(
+                        c =>
+                        {
+                            var invalidDataTypeCreation = symbolsWhereTypeIsCreated
+                                .Where(x => !symbolsWhereLocaleIsSet.Contains(x.Symbol));
+
+                            foreach (var invalidCreation in invalidDataTypeCreation)
+                            {
+                                var typeName = GetConstructedTypeName(invalidCreation.Symbol);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: You could use `GetSymbolType(invalidCreation.Symbol)?.Name` instead (and remove your method)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+                                if (typeName == null)
+                                {
+                                    continue;
+                                }
+
+                                c.ReportDiagnosticIfNonGenerated(Diagnostic.Create(rule,
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: Is it required to use *IfNonGenerated? The collected data comes from non-generated code, this seems redundant

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+                            var variableSyntax = GetAssignedToSyntax(objectCreation);
+                            if (variableSyntax == null)
+                            {
+                                return;
+                            }
+
+                            var variableSymbol = variableSyntax is IdentifierNameSyntax
+                                ? c.SemanticModel.GetSymbolInfo(variableSyntax).Symbol
+                                : c.SemanticModel.GetDeclaredSymbol(variableSyntax);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: Do you need both GetSymbolInfo and GetDeclaredSymbol?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 10 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: You are right, this is not needed.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 11 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: @valhristov Some cases don't work with `GetSymbolInfo` others don't with `GetDeclaredSymbol` but if you have a nice suggestion, I am listening.

Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: I was curious if this is a mistake, because we talked about it and I was left with the impression that only one is needed in this case. But if it isn't, I am fine with the solution.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 9 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: invalidCreation.Symbol.GetSymbolType()?.Name

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.cs (hunk 5 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.cs]: The check to the culture is not required by this rule (maybe we could extend at some point).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Vsix/source.extension.vsixmanifest (hunk 2 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Vsix/source.extension.vsixmanifest]: I can't manage to undo anything on this :C

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+            if (memberAccess != null)
+            {
+                return memberAccess.Expression != null
+                    ? model.GetSymbolInfo(memberAccess.Expression).Symbol
+                    : null;
+            }
+
+            var conditionalAccess = variable as ConditionalAccessExpressionSyntax;
+            return conditionalAccess?.Expression == null
+                ? model.GetSymbolInfo(conditionalAccess.Expression).Symbol
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: MAJOR 'conditionalAccess' is null on at least one execution path. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2259)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> checkedTypes = new HashSet<KnownType>
+        {
+            KnownType.System_Data_DataTable,
+            KnownType.System_Data_DataSet
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+                    : null;
+            }
+
+            var conditionalAccess = variable as ConditionalAccessExpressionSyntax;
+            return conditionalAccess?.Expression == null
+                ? model.GetSymbolInfo(conditionalAccess.Expression).Symbol
+                : null;
+        }
+
+        private static string GetConstructedTypeName(ISymbol symbol)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: MAJOR Remove the unused private method 'GetConstructedTypeName'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Vsix/source.extension.vsixmanifest (hunk 2 above)
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Vsix/source.extension.vsixmanifest]: Why not? Is it automatically regenerated by VS?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 12 above)
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: We definitely need a test for this case.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> checkedTypes = new HashSet<KnownType>
+        {
+            KnownType.System_Data_DataTable,
+            KnownType.System_Data_DataSet
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+            }
+
+            var memberAccess = variable as MemberAccessExpressionSyntax;
+
+            return memberAccess?.Expression != null
+                ? model.GetSymbolInfo(memberAccess.Expression).Symbol
+                : null;
+        }
+
+        private static string GetConstructedTypeName(ISymbol symbol)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: MAJOR Remove the unused private method 'GetConstructedTypeName'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
